helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -1285,4 +1266,85 @@ public void handleSessionEstablishmentError(Throwable error) throws Exception {
+        new RealmAwareZkClient.RealmAwareZkClientConfig();
+
+    clientConfig.setZkSerializer(zkSerializer)
+        .setConnectInitTimeout(_connectionInitTimeout)
+        .setMonitorType(_instanceType.name())
+        .setMonitorKey(_clusterName)
+        .setMonitorInstanceName(_instanceName)
+        .setMonitorRootPathOnly(isMonitorRootPathOnly());
+
+    RealmAwareZkClient newClient;
+    if (_instanceType == InstanceType.ADMINISTRATOR) {
+      newClient = buildSharedZkClient(connectionConfig, clientConfig);
+    } else {
+      newClient = buildDedicatedZkClient(connectionConfig, clientConfig);
+    }
+
+    return newClient;
+  }
+
+  private RealmAwareZkClient buildSharedZkClient(
+      RealmAwareZkClient.RealmAwareZkConnectionConfig connectionConfig,
+      RealmAwareZkClient.RealmAwareZkClientConfig clientConfig) {
+    RealmAwareZkClient zkClient;
+    try {
+      zkClient =
+          SharedZkClientFactory.getInstance().buildZkClient(connectionConfig, clientConfig);
+    } catch (IllegalArgumentException | IOException | InvalidRoutingDataException e) {
+      LOG.warn("Not able to build realm-aware shared ZK client for sharding key: {}, caused by: {}"
+              + " Fallback to HelixZkClient.", connectionConfig.getZkRealmShardingKey(),
+          e.getMessage());
+
+      // Fallback to HelixZkClient
+      HelixZkClient.ZkConnectionConfig helixZkConnectionConfig =
+          new HelixZkClient.ZkConnectionConfig(_zkAddress);
+      helixZkConnectionConfig.setSessionTimeout(connectionConfig.getSessionTimeout());
+      zkClient = SharedZkClientFactory.getInstance()
+          .buildZkClient(helixZkConnectionConfig, clientConfig.createHelixZkClientConfig());
+    }
+
+    return zkClient;
+  }
+
+  private RealmAwareZkClient buildDedicatedZkClient(
+      RealmAwareZkClient.RealmAwareZkConnectionConfig connectionConfig,
+      RealmAwareZkClient.RealmAwareZkClientConfig clientConfig) {
+    RealmAwareZkClient zkClient;
+    try {
+      zkClient =
+          DedicatedZkClientFactory.getInstance().buildZkClient(connectionConfig, clientConfig);
+    } catch (IllegalArgumentException | IOException | InvalidRoutingDataException e) {
+      LOG.warn("Not able to build realm-aware Dedicated ZK client for sharding key: {}, caused by: {}"
+              + " Fall back to HelixZkClient.", connectionConfig.getZkRealmShardingKey(),
+          e.getMessage());
+
+      // Fallback to HelixZkClient
+      HelixZkClient.ZkConnectionConfig helixZkConnectionConfig =
+          new HelixZkClient.ZkConnectionConfig(_zkAddress);
+      helixZkConnectionConfig.setSessionTimeout(connectionConfig.getSessionTimeout());
+      zkClient = DedicatedZkClientFactory.getInstance()
+          .buildZkClient(helixZkConnectionConfig, clientConfig.createHelixZkClientConfig());
+    }
+
+    return zkClient;
+  }
+
+  private String buildShardingKey() {
+    return _clusterName.charAt(0) == '/' ? _clusterName : SLASH + _clusterName;
```

Comment:
Nit: not much point declaring `SLASH` if it's character counter-part is a local variable.